R/data.table.R

```diff
@@ -337,6 +337,30 @@ replace_dot_alias = function(e) {
         isub = isub[[2L]]
     }
 
+    if (is.call(isub) && as.character(isub[[1]])[[1L]] %chin% c('top.n', 'filter.at')) {  #TODO , 'sample.n'
+      if (verbose) cat('Inside subset.i helper functions. The following will be evaluated\n')
+      l <- eval(isub)
+      if (verbose) print(l)
+      
+      if (is.null(names(l))){ #unnamed lists need to use eval
+        if (length(l) == 1) {
+          eval(l[[1]], parent.frame(), parent.frame())
+        } else {
+          for(itm in l) {eval(itm)}
+        }
+        isub = ind
+      } else if (!is.null(l$f_order)){
+        #next three lines copied from elsewhere except isub is replaced with l[['i']]
+        ienv = new.env(parent=parent.frame())
+        if (getOption("datatable.optimize")>=1L) assign("order", forder, ienv)
+        isub = tryCatch(eval(.massagei(l[['i']]), x, ienv), error=function(e) .checkTypos(e, names_x))
+      } else if (is.null(l$by)) {
+        isub = do.call(`[.data.table`, l)
```

Comment:
This is generally a bad idea to call `[.data.table` from inside `[.data.table`. I recall there was a single(?) place where we already did that, and we wanted to get rid of it. If there is another way, then we should consider to use that instead.

Reply:
line 2003 - ```i = x[i, which=TRUE]```. I can untangle the calls without ```by``` - just need to include extra code to handle the ```.SDcols``` argument. I would need to dig very deeply to remove the call with the ```by``` but I will look into it. It would be easier if https://github.com/Rdatatable/data.table/issues/852 were further along to make the various ```by``` and ```j``` checks more modular.

Reply:
@jangorecki I have taken this comment under consideration. I am starting to look into C methods to do a quicker ```topn``` which should assist me with ```filter.at```. Plus, the former is an actual feature that has upvotes. 

For example, using ```forderv``` can get head by group with:
```
dt = data.table(V1 = sample(50, 500, TRUE))
tmp = data.table:::forderv(dt[['V1']], sort = FALSE, retGrp = T)
dt[tmp[attr(tmp, 'starts')]]
```
But obviously, we allocate a large integer vector just to immediately subset. A new method that only returns a vector of what is used should be more performant.

Also, I appreciate your time and comments. I will likely close this is in a couple of days as while this is speedy as is, I acknowledge it is somewhat hacky.